src/DescriptorsCommand.cc

```diff
@@ -409,94 +412,240 @@ void AddDescriptor::retrieve_aws_descriptorSet(const std::string &set_path) {
+    // if(_use_aws_storage)
+    // {
+    //     //delete files in set_path
+    //     std::uintmax_t n = fs::remove_all(set_path);
+    //     std::cout << "Deleted " << n << " files or directories\n";
+    // }
+
+    return 0;
 
+}
+
+int AddDescriptor::add_descriptor_batch(PMGDQuery &query,
+                                         const Json::Value &jsoncmd,
+                                         const std::string &blob, int grp_id,
+                                         Json::Value &error){
+
+    int expected_blb_size;
+    int nr_expected_descs;
+    int dimensions;
+
+    //Extract set name
+    const Json::Value &cmd = jsoncmd[_cmd_name];
+    const std::string set_name = cmd["set"].asString();
+
+    Json::Value props = get_value<Json::Value>(cmd, "properties");
+
+    //extract properties list and get filepath/object location of set
+    Json::Value prop_list = get_value<Json::Value>(cmd, "batch_properties");
+    const std::string set_path = get_set_path(query, set_name, dimensions);
+
+    if (set_path.empty()) {
+        error["info"] = "Set " + set_name + " not found";
+        error["status"] = RSCommand::Error;
+        return -1;
+    }
+
+    std::string label = get_value<std::string>(cmd, "label", "None");
+    props[VDMS_DESC_LABEL_PROP] = label;
+
+    // retrieve the descriptor set from AWS here
+    // operations are currently done in memory with no subsequent write to disk
+    // so there's no need to re-upload to AWS
     if (_use_aws_storage) {
-      // delete files in set_path
-      std::uintmax_t n = fs::remove_all(set_path);
-      std::cout << "Deleted " << n << " files or directories\n";
+        retrieve_aws_descriptorSet(set_path);
     }
 
-    return -1;
-  }
+    // Note dimensionse are based on a 32 bit integer, hence the /4 math on size
+    // as the string blob is sized in 8 bit ints.
+    nr_expected_descs = prop_list.size();
+    expected_blb_size = nr_expected_descs * dimensions * 4;
+
+    //Verify length of input is matching expectations
+    if (blob.length() != expected_blb_size) {
+        std::cerr << "AddDescriptor::insert_descriptor: ";
+        std::cerr << "Expectected Blob Length Does Not Match Input ";
+        std::cerr << blob.length() << " != " << expected_blb_size << std::endl;
+        error["info"] = "FV Input Length Mismatch";
+        return -1;
+    }
 
-  props[VDMS_DESC_ID_PROP] = Json::Int64(id);
+    //TODO modify insert descriptor to handle batches
+    long id = insert_descriptor(blob, set_path, nr_expected_descs, label, error);
 
-  int node_ref = get_value<int>(cmd, "_ref", query.get_available_reference());
+    if (id < 0) {
+        error["status"] = RSCommand::Error;
 
-  query.AddNode(node_ref, VDMS_DESC_TAG, props, Json::nullValue);
+        if (_use_aws_storage) {
+            // delete files in set_path
+            std::uintmax_t n = fs::remove_all(set_path);
+            std::cout << "Deleted " << n << " files or directories\n";
```

Comment:
Why do we want to delete all existing files for the descriptor set when an insert of a new descriptor fails?

Reply:
This is some weird hold-over logic from the iterative implementation. My understanding is because they're staging local files before popping them over to the S3 target. I'm hesitant about nuking this cause I don't have enough insight yet to exactly _why_ they took this approach.